demos/eigenvalues_QG_basinmodes/qgbasinmodes.rst

```diff
@@ -0,0 +1,209 @@
+reside. ::
+
+   Vcg  = FunctionSpace(mesh,'CG',3)
+
+We impose zero Dirichlet boundary conditions, in a strong sense, which
+guarantee that we have no-normal flow at the boundary walls. ::
+
+   bc = DirichletBC(Vcg, 0.0, (1, 2, 3, 4))
+
+The two non-dimensinal parameters are the :math:`\beta` parameter, set
+by the sphericity of the Earth, and the Froude number, the relative
+importance of rotation to stratification. ::
+
+   beta = Constant('1.0')
+   F    = Constant('1.0')
+
+Additionally, we can create some Functions and arrays to store our
+solutions. ::
+
+   eigenmodes_real, eigenmodes_imag = Function(Vcg), Function(Vcg)
+   eigenvaluer, eigenvaluei = [], []
+
+We define the Test Function :math:`\phi` and the Trial Function
+:math:`\psi` in our function space. ::
+
+   phi, psi = TestFunction(Vcg), TrialFunction(Vcg)
+
+To build the weak formulation of our equation we need to build two PETSc
+matrices in the form of a generalized eigenvalue problem,
+:math:`A\psi = \lambda M\psi`. We impose the boundary conditions on the
+mass matrix :math:`m`, since that is where we used integration by parts. ::
+
+   a =  beta*phi*psi.dx(0)*dx
+   m = -inner(grad(psi), grad(phi))*dx - F*psi*phi*dx
+   petsc_a = assemble(a).M.handle
+   petsc_m = assemble(m, bcs=bc).M.handle
+
+We can declare how many eigenpairs, eigenfunctions and eigenvalues, we
+want to find ::
+
+   num_eigenvalues = 2
+
+Next we will impose parameters onto our eigenvalue solver. The first is
+specifying that we have an generalized eigenvalue problem that is
+nonhermitian. The second specifies the spectral transform shift factor
+to be non-zero. The third requires we use a Krylov-Schur method,
+which is the default so this is not strictly necessary. Then, we ask for
+the eigenvalues with the largest imaginary part. Finally, we specify the
+tolerance. ::
+
+   opts = PETSc.Options
+   opts.setValue("eps_gen_non_hermitian", None)
+   opts.setValue("st_pc_factor_shift_type", "NONZERO")
+   opts.setValue("eps_type", "krylovschur")
+   opts.setValue("eps_largest_imaginary", None)
+   opts.setValue("eps_tol", 1e-10)
+
+Finally, we build our eigenvalue solver using SLEPc. We add our PETSc
+matrices into the solver as operators and use setFromOptions() to call
+the PETSc parameters we previously declared. ::
+
+   es = SLEPc.EPS().create(comm=COMM_WORLD)
+   es.setDimensions(num_eigenvalues)
+   es.setOperators(petsc_a, petsc_m)
+   es.setFromOptions()
+   es.solve()
+
+Additionally we can find the number of converged eigenvalues. ::
+
+   nconv = es.getConverged()
+   print nconv
+
+In the range of all converged eigenpairs (nconv), we can retrieve all
+real and imaginary parts of the eigenvalues.  ::
+
+   vr, vi = petsc_a.getVecs()
+   for i in range(nconv):
+       lam = es.getEigenpair(i, vr, vi)
+       eigenvaluer.append(lam.real)
+       eigenvaluei.append(lam.imag)
+
+and the eigenfunctions ::
+
+   eigenmodes_real.vector()[:], eigenmodes_imag.vector()[:] = vr, vi
```

Comment:
This bit of code doesn't quite do what you expect?  Let's say we converge two eigenvalues.  We collect the real and imaginary parts of both eigenvalues, but only save the real and imaginary parts of the eigen*vector* for the second of the two converged eigenvalues.